contracts/extensions/ReputationBootstrapper.sol

```diff
@@ -0,0 +1,219 @@
+
+    colony = IColony(_colony);
+    token = colony.getToken();
+
+    address colonyNetwork = colony.getColonyNetwork();
+    address repCycle = IColonyNetwork(colonyNetwork).getReputationMiningCycle(false);
+    decayPeriod = IReputationMiningCycle(repCycle).getMiningWindowDuration();
+    (decayNumerator, decayDenominator) = IReputationMiningCycle(repCycle).getDecayConstant();
+  }
+
+  /// @notice Called when upgrading the extension
+  function finishUpgrade() public override auth {}
+
+  /// @notice Called when deprecating (or undeprecating) the extension
+  function deprecate(bool _deprecated) public override auth {
+    deprecated = _deprecated;
+  }
+
+  /// @notice Called when uninstalling the extension
+  function uninstall() public override auth {
+    uint256 balance = ERC20(token).balanceOf(address(this));
+    require(ERC20(token).transfer(address(colony), balance), "reputation-bootstrapper-transfer-failed");
+
+    selfdestruct(address(uint160(address(colony))));
+  }
+
+  // Public
+
+  /// @notice Set an arbitrary number of grants
+  /// @param _paid An array of booleans indicated pair or unpaid
+  /// @param _hashedSecrets An array of (hashed) secrets
+  /// @param _amounts An array of reputation amounts claimable by the secret
+  function setGrants(bool[] memory _paid, bytes32[] memory _hashedSecrets, uint256[] memory _amounts) public onlyRoot notDeprecated {
+    require(_paid.length == _hashedSecrets.length, "reputation-bootstrapper-invalid-arguments");
+    require(_hashedSecrets.length == _amounts.length, "reputation-bootstrapper-invalid-arguments");
+    uint256 balance = ERC20(token).balanceOf(address(this));
+
+    for (uint256 i; i < _hashedSecrets.length; i++) {
+      require(_amounts[i] <= INT128_MAX, "reputation-bootstrapper-invalid-amount");
+
+      if (_paid[i]) {
+        uint256 priorGrant = grants[_paid[i]][_hashedSecrets[i]].amount;
+        totalPayableGrants = totalPayableGrants - priorGrant + _amounts[i];
+      }
+
+      grants[_paid[i]][_hashedSecrets[i]] = Grant(_amounts[i], block.timestamp);
+
+      emit GrantSet(_paid[i], _hashedSecrets[i], _amounts[i]);
+    }
+
+    require(totalPayableGrants <= balance, "reputation-bootstrapper-insufficient-balance");
+  }
+
+  /// @notice Commit the secret, beginning the security delay window
+  /// @param _committedSecret A sha256 hash of (userAddress, secret)
+  function commitSecret(bytes32 _committedSecret) public notDeprecated {
+    bytes32 addressHash = keccak256(abi.encodePacked(msgSender(), _committedSecret));
+    committedSecrets[addressHash] = block.timestamp;
+  }
+
+  /// @notice Claim the grant, after committing the secret and having the security delay elapse
+  /// @param _secret The secret corresponding to a reputation grant
+  function claimGrant(bool _paid, uint256 _secret) public notDeprecated {
+    bytes32 committedSecret = keccak256(abi.encodePacked(msgSender(), _secret));
```

Comment:
Is this not still an unnecessary amount of hashing? This could just be `keccak256(abi.encodePacked(_secret))`, no?

Reply:
I don't think so, because otherwise someone could supply a hash on behalf of another user. Although I suppose there's nothing wrong with that because it is only to the user's advantage (e.g. they can claim earlier than they otherwise would have)...